src/Task.ts

```diff
@@ -862,11 +862,37 @@ export class Task {
         )})`;
     }
 
+    /*
+     * Escape a string so it can be used as part of a RegExp literally.
+     * Taken from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#escaping
+     */
+    private escapeRegExp(s: string) {
+        return s.replace(/([.*+?^=!:${}()|[]\/\\])/g, '\\$1');
```

Comment:
I tried understanding this regex, and was not able to, from my current knowledge of regular expressions.

So I tried the expression in https://regex101.com and couldn't get it to match individual characters. Which may be because I chose the wrong regex engine.

So I did what I always do in such situations - try to break the code.

I changed `escapeRegExp()` so it returned `s`. And the `Task.test.ts` tests all passed.

I may be missing something, but it seems as though there is no test that exercises `escapeRegExp()`?

Reply:
@claremacrae The purpose of this Regex seems to be to add the escape char `\` to any character in the string that needs escaping when used in a regex. So I think the best way to test this one would be to try a string/global filter with any (or all?) of these chars in it: .`*` `+` `?` `^` `=` `!` `:` `$` `{` `}` `(` `)` `|` `[` `]` `/` or `\`

If it's not escaping properly, it likely won't match the global filter _or_ I guess it could match more than just the global filter... ?

As @AnnaKornfeldSimpson states, I do believe the regex is missing a `\`. Of course, I have had lint correct me for extra escapes in my regex before so I could be wrong...

**Doesn't match all the chars**
![image](https://user-images.githubusercontent.com/8586879/180890269-2bbc80c6-677d-4510-af50-caa1a9b64283.png)

**Matches all the chars**
![image](https://user-images.githubusercontent.com/8586879/180890340-7f7480e2-f437-40d9-bc5c-768c33f5d917.png)

Reply:
The goal of `getDescriptionWithoutGlobalFilter` is to remove the global filter only if it is a complete word, and there's a regex to test that (`globalFilterRegex`). However, a user-entered string like `globalFilter` can't be used as part of a regex as it is, because it might contain special characters that will be interpreted as RegEx language, e.g. if the global filter is `#t[xy]` and it's used in the regex as it is, it would match `#tx` and `#ty` instead of what we want it to...
That's where escaping comes in, the function `escapeRegExp` takes a string that could contain special characters and replaces them with the same character preceded by a slash. So `#t[xy]` would enter our regex as `#t/[xy/]` and the brackets would match literally.
I did not make it up, it's taken almost as-is from the formal MDN.
I say "almost" because @robosquiggles noticed correctly that there are two slashes "missing" in my version; that's because the linter configured in Tasks outputs an error for the original expression that the slashes of `[` and `]` are redundant. That's a matter of Regex flavor I guess.

Reply:
---
creation date: 2022-07-26 12:05
tags: 
aliases: []
---

# delete me

Thanks for all the replies - super helpful!

This is long, as it turns out that there is a lot to say about that one little regex in `escapeRegExp()`...

## Summary

- The regex in `escapeRegExp()` is indeed broken, with a missing `\`, which prevents it escaping any characters
- Some of the characters it escapes result in an error message  when used, such as:
    - `Invalid regular expression: /(^|\s)hello\=world($|\s)/: Invalid escape`
- Writing tests before adding each of the characters supposed to be escaped revealed much useful info
- Over to @esm7 to decide how to move forward

## Clarification of understanding

> I tried understanding this regex, and was not able to, from my current knowledge of regular expressions.

Yeah, sorry, I was unclear. What I meant by this was that I knew the intent of the code, from the comments and the linked article.

However when I tried by eye to read all the characters in the regex, with some escaping, I was not confident that I could precisely say which bits of the regex were searching for which characters.

In my experience, regex that are not immediately glanceable to understand them need to:

- either have incredibly robust tests
- or to be broken down in to component parts, with a comment against each component saying the purpose (e.g. 'escape any [ chars')

Otherwise:

- they may well currently have a bug
- if they don't have a bug now, there's a fair chance that someone in future will accidentally introduce one

This was all before I saw that @AnnaKornfeldSimpson had spotted an error via regex101.com and @robosquiggles had confirmed.

## Experiment 1 - do the current tests break without escaping the global filter?

If the regex in `escapeRegExp()` was being covered by any tests, and I removed its escaping, I would get a test failure.

With this change, as I mentioned earlier, all of the tests pass:

```diff
     private escapeRegExp(s: string) {
-        return s.replace(/([.*+?^=!:${}()|[]\/\\])/g, '\\$1');
+        // return s.replace(/([.*+?^=!:${}()|[]\/\\])/g, '\\$1');
+        return s;
     }
``` 

So that told me that the tests cannot possibly be using any inputs that depend on any of the characters that are meant to be escaped.

## Experiment 2 - try to add a failing test for one of the special characters - and fix it.

<!--
starting point 28b7f5a97d77d604996c92139886ec687ed23fc3
-->

<!--
- Last commit:  14dadeb98b76fc51bdd1b7bba9d49dde78263480
-->

**Adding a test:**

I was able to add a failing test that used  `*` as the global filter:

https://github.com/claremacrae/obsidian-tasks/blob/14dadeb98b76fc51bdd1b7bba9d49dde78263480/tests/Task.test.ts#L996-L1055

**Fixing the regular expression:**

- Once I had the failing test, I was then able to fix the regex, adding the missing `\` before the `[`.
- WebStorm then told me about a redundant `\`. which turned out to be in this: `\/`.
- I speculate that there was previously a warning about a redundant `\` and it was silenced by removing the wrong `\`.
- Removing that gave me this fix: [fix: Fix failing test - asterisk now escaped](https://github.com/obsidian-tasks-group/obsidian-tasks/commit/14dadeb98b76fc51bdd1b7bba9d49dde78263480)

**The changes:**

- This was all done on my branch:
    - [fix-global-filter-removal-fix-regex](https://github.com/claremacrae/obsidian-tasks/tree/fix-global-filter-removal-fix-regex)
- Here are all the changes for this experiment:
    - [Comparing obsidian-tasks-group:28b7f5a97d77d604996c92139886ec687ed23fc3...claremacrae:fix-global-filter-removal-fix-regex](https://github.com/obsidian-tasks-group/obsidian-tasks/compare/28b7f5a97d77d604996c92139886ec687ed23fc3...claremacrae:obsidian-tasks:fix-global-filter-removal-fix-regex)

## Experiment 3 - try TDD-ing it

My immediate thought after **Experiment 1** was:

> This is where TDD shines... if the regex started empty, and then a failing test added for each character it needed to detect, the regex would be known to be correct, and protected for the future.

So now I thought I would give it a go, on the branch [fix-global-filter-removal-tdd](https://github.com/claremacrae/obsidian-tasks/tree/fix-global-filter-removal-tdd), based off the previous section, but before the I fixed the regex.

<!--
- Last commit: 13bc7321460faece2674f1d6521c3a33a1347737
- -->
Here are all the changes on the branch, including the addition of tests shown in the previous section:

- [Comparing obsidian-tasks-group:28b7f5a97d77d604996c92139886ec687ed23fc3...claremacrae:fix-global-filter-removal-tdd](https://github.com/obsidian-tasks-group/obsidian-tasks/compare/28b7f5a97d77d604996c92139886ec687ed23fc3...claremacrae:obsidian-tasks:fix-global-filter-removal-tdd)

Here are just the commits that build up the regular expression in `Task.escapeRegExp()` on character at a time.

- [Comparing obsidian-tasks-group:c26aa6f7f6a4a22a02dc36140e01b3a34ae1b895...claremacrae:fix-global-filter-removal-tdd](https://github.com/obsidian-tasks-group/obsidian-tasks/compare/c26aa6f7f6a4a22a02dc36140e01b3a34ae1b895...claremacrae:obsidian-tasks:fix-global-filter-removal-tdd)
- The changes are best reviewed one at a time.

Takeaways for this implementation of  `escapeRegExp()` are:

- Making the regex escape `=`, `!` and `:` seems to make the replacement fail, complaining about invalid escaping
- I also wasn't able to make a failing test to show that `/` needed escaping
- Whatever characters are escaped by `escapeRegExp()`, we definitely need to have tests that execute the escaping for every one of those characters, to ensure that they leave a valid regex.

## What next?

I hope this gives some kind of idea about why I am cautious about regular expressions!

Anyway, I think my only concrete conclusions from this are:

- `escapeRegExp()` can't be released in its current form.
- there needs to be tests that exercise thoroughly every single one of the characters that are supposed to be being escaped.

What I don't know is:

- whether the fix is to stick with the removal of the escaping of characters that gave error messages.
- or whether to make `escapeRegExp()` more aware of certain regex patterns.

I considered changing the Add/Edit modal to ban certain characters from global filters, but with over 100,000 downloads (!) there's bound to be someone using one of each them already.

Over to you, @esm7, to decide what to do next!

Reply:
See my comments above, but just to pick out one particular comment:

> I say "almost" because @robosquiggles noticed correctly that there are two slashes "missing" in my version; that's because the linter configured in Tasks outputs an error for the original expression that the slashes of `[` and `]` are redundant. That's a matter of Regex flavor I guess.

The linter was correct. The wrong `\` was removed, which silenced the linter but broke the code.

Reply:
I'm resolving this having worked out that I can merge this PR and then make the additions to the tests separately. Sorry for not working that out from the start,